### ISSUE TITLE
internal/shader: utility functions to treat normalized values

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#2644]
### Operating System

- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] FreeBSD
- [ ] OpenBSD
- [ ] Android
- [ ] iOS
- [ ] Nintendo Switch
- [ ] Xbox
- [ ] Web Browsers

### What feature would you like to be added?

I propose to add these functions

* `func normalizeDstPos(position vec2) vec2`: converts `position` in pixels to [0, 1] of the destination image.
* `func normalizeSrcPos(position vec2) vec2`: converts `position` in texles to [0, 1] of the source image.

TBD: Do we need inverted functions?

### Why is this needed?

From the discussion of #2639, we found that normalizing positions of textures to [0, 1] is very common. To do this, users need to understand how images and textures work in Ebitengine, and to write some tricky code with an origin and a texture. Thus, it should be very useful to have utility functions to do that for Kage.

For more details, see https://docs.google.com/document/d/1lJTKxMqyyO-LxNS4KmuNakQwlhdBgzkmP2LNwpJk5bg/edit?usp=sharing

**Comment by hajimehoshi** [comment]
For #1870, we might have to have these functions:

* `func normalizeDstPosition(pos vec2) vec2`: converts pos in pixels to [0, 1] of the destination image.
* `func normalizeSrcNPosition(pos vec2) vec2`: converts pos in texles to [0, 1] of the source image. N is 0 to 3.

**Comment by tinne26** [comment]
In my opinion, having the destination position in pixels in the natural destination range (instead of normalized to [0, 1]) is also useful and should be considered:
```Golang
func localizeDstPosition(pos vec2) vec2 { // localize, extract, transport, transform, normalize?
    orig, size := imageDstRegionOnTexture()
    return (pos - orig)/size
}
```
As an example, all kage-desk tutorials use destination positions like this and should a normalization function like this one to be correct on >=v2.5.0.

I find the terms "unit" or "unit range" useful when working with values in the [0, 1] range. "Normalized" also refers to the same concept, but in this situation it can be confusing and ambiguous. We have the "raw" coordinates on atlases and regions, which then we can "transform" or "normalize" to the natural ranges that the user would expect, which then we can further "normalize" or remap to the unit range. They are all types of coordinate transformations, but I think the terminology, as we accept that atlases and atlas regions are relevant concepts exposed to the users when working with shaders, could be better standardized.

**Comment by tinne26** [comment]
By the way, to use `position` directly in pixels, without normalizing to [0, 1], don't we also need to use the origin from `imageDstRegionOnTexture()` and transform it from texels to pixels with tricky calculations too? Because if this is the case, we should also have a normalization function for it. Until we actually get the pixels mode, this specific conversion (if I'm imagining it correctly) seems extremely convoluted.

**Comment by hajimehoshi** [comment]
> don't we also need to use the origin from imageDstRegionOnTexture() and transform it from texels to pixels with tricky calculations too?

In the pixel mode, we don't treat texels even for destination textures, right?

**Comment by tinne26** [comment]
Right. What I'm asking is this: given a target of 400x400, will `position` go from `0.5` to `395.5`, or do we have to correct first for the destination origin? Unless I'm misunderstanding something (which is highly likely), I assume that in pixel mode we simply would do:
```Golang
//kage:pixel-mode
func Fragment(position vec4, _ vec2, _ vec4) vec4 {
    origin, _ := imageDstRegionOnTexture()
    dstPosition := position.xy - origin
}
```
But without pixel mode, this is currently much harder to do? Or is this step already unnecessary?

**Comment by hajimehoshi** [comment]
> But without pixel mode, this is currently much harder to do? Or is this step already unnecessary?

If you are talking about how to get [0, 1] value, yes, this is a little much harder, but it's still possible. See https://github.com/hajimehoshi/ebiten/blob/2.5/examples/shader/texel.go (texel mode) vs https://github.com/hajimehoshi/ebiten/blob/main/examples/shader/texel.go (pixel mode).

> I know that for many subimages position won't start at zero, but I still don't know if destination images with (0, 0) origin require atlas origin normalization or not to use position in the "natural range" of the destination image. If doing this is necessary, maybe we should have a normalization function for it too?

For a sub-image, I think we should provide a function to return [0, 1] assuming the sub-image's left-upper is (0, 0) and right-bottom is (1, 1). But this might be ambiguous. Let me think.

**Comment by tinne26** [comment]
No, I'm talking about how to get the `position` value in `[0, N]` where `N` is the size of the destination image (let's assume it's a square image), without considering special cases like subimages or scaling due to anti-aliasing. Technically it's `[0.5, N - 0.5]`, but you get the point. I don't know if we need to account for the origin or not in the current state.

**Comment by hajimehoshi** [comment]
This requires more discussion. I'll reset the milestone.
